Add a test for TLS pipelining

mattcaswell opened the pull request:
TLS pipelining provides the ability for libssl to read or write multiple records in parallel. It requires special ciphers to do this, and there are currently no built-in ciphers that provide this capability. However, the dasync engine does have such a cipher, so we add a test for this capability using that engine.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated

mattcaswell commented:
This went in surprisingly smoothly and easily. Given the length of time this code has been there without a test, and the recent refactor of the record layer - I was expecting problems.

mattcaswell commented:
> The address sanitizer disagrees

Should be fixed now.

t8m reviewed (approved):
Approved with the typos fixed.

mattcaswell commented:
Ping @openssl/committers for second review?

tmshort commented on test/sslapitest.c:
```diff
@@ -10150,6 +10151,143 @@ static int test_tls13_record_padding(int idx)
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL, *peera, *peerb;
+    int testresult = 0, numreads;
+    char *msg = "A test message that must be at least 50 bytes long";
```
Cute, but perhaps something a little less cute, and a bit more obvious as to the length (yes, I counted it).

mattcaswell replied:
Done

tmshort reviewed (commented):
Mostly just a question on the use of ENGINE in the master branch, and whether this testing is sufficient (given the small data size).

mattcaswell commented on test/sslapitest.c:
```diff
@@ -10150,6 +10151,143 @@ static int test_tls13_record_padding(int idx)
+    char buf[80];
+    ENGINE *e;
+
+    if (!TEST_ptr(e = ENGINE_by_id("dasync")))
```
> Are they fully deprecated

Yes

> Should this not be a provider or something else?

Yes, really this test *should* be using a provider. Unfortunately at the moment we are in the slightly strange situation that libssl pipeline support is not deprecated, but in order to use that support you must use a pipeline capable cipher - and we have not added the support for pipeline capable ciphers into providers. So the *only* way you can possibly use the non-deprecated libssl pipeline support is by using a deprecated ENGINE. I consider this an accidental omission which we should fix. But that goes way beyond what I was planning to do with this PR, i.e. just make sure what we've currently got isn't broken.

mattcaswell replied:
See also #18298

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -10150,6 +10151,143 @@ static int test_tls13_record_padding(int idx)
+    /*
+     * peera is always configured for pipelining, while peerb is not. We have 5
+     * pipelines with 10 bytes per pipeline (50 bytes in total). This is a
+     * ridiculously small number of bytes - but sufficient for our purposes
```
I've added some more testing including one for large (maximal) pipelines.

mattcaswell commented:
Adding the additional testing to this as suggested by @tmshort was a worthwhile exercise since it highlighted a couple of problems in the read pipelining implementation. I've fixed those issues and also made a tweak to the documentation as a result.

I suspect that the issues I fixed predated the record layer refactor so its possible those issues exist in 3.0 and 1.1.1 - but I've not investigated that.

mattcaswell commented:
This now needs re-review since there have been some significant changes since the original version.

t8m commented on ssl/record/methods/tls_common.c:
```diff
@@ -286,24 +292,6 @@ int tls_default_read_n(OSSL_RECORD_LAYER *rl, size_t n, size_t max, int extend,
         /* start with empty packet ... */
         if (left == 0) {
```
You can now drop the `{}`

t8m commented on doc/man3/SSL_CTX_set_split_send_fragment.pod:
```diff
@@ -96,7 +96,10 @@ into the buffer. Without this set data is read into the read buffer one record
-function (see L<SSL_pending(3)>).
+function (see L<SSL_pending(3)>). In addition the default size of the internal
+read buffer is multiplied by the number of pipelines available to ensure that we
+can read multiple records in one go. This can therefore have a signficiant
```
Typo: significant

tmshort commented on test/sslapitest.c:
```diff
@@ -10164,15 +10164,21 @@ static int test_tls13_record_padding(int idx)
-    char buf[80];
+    /* A 55 byte message */
+    unsigned char *msg = (unsigned char *)
+        "0123456789012345678901234567890123456789012345678901234";
```
"0-9A-Za-z" is 62 characters, perhaps?

mattcaswell commented:
All feedback addressed. @t8m - please take another look.

mattcaswell commented:
Ping for second review @openssl/committers

tmshort commented on ssl/record/methods/tls1_meth.c:
```diff
@@ -162,6 +162,7 @@ static int tls1_cipher(OSSL_RECORD_LAYER *rl, SSL3_RECORD *recs, size_t n_recs,
     EVP_CIPHER_CTX *ds;
     size_t reclen[SSL_MAX_PIPELINES];
     unsigned char buf[SSL_MAX_PIPELINES][EVP_AEAD_TLS1_AAD_LEN];
+    unsigned char *data[SSL_MAX_PIPELINES];
```
Why was this moved to the outer scope? It's the only change in this function.

mattcaswell replied:
This was due to use-after-scope error picked up by asan. The buffers set by `EVP_CTRL_SET_PIPELINE_OUTPUT_BUFS` must still be valid at the point that we call `EVP_CipherUpdate` - which was not the case when this was in its original location.

tmshort commented on ssl/record/methods/tls_common.c:
```diff
@@ -284,27 +290,9 @@ int tls_default_read_n(OSSL_RECORD_LAYER *rl, size_t n, size_t max, int extend,
-        if (left == 0) {
+        if (left == 0)
             rb->offset = align;
-        } else if (align != 0 && left >= SSL3_RT_HEADER_LENGTH) {
-            /*
-             * check if next packet length is large enough to justify payload
-             * alignment...
-             */
-            pkt = rb->buf + rb->offset;
-            if (pkt[0] == SSL3_RT_APPLICATION_DATA
-                    && (pkt[3] << 8 | pkt[4]) >= 128) {
-                /*
-                 * Note that even if packet is corrupted and its length field
-                 * is insane, we can only be led to wrong decision about
-                 * whether memmove will occur or not. Header values has no
-                 * effect on memmove arguments and therefore no buffer
-                 * overrun can be triggered.
-                 */
-                memmove(rb->buf + align, pkt, left);
-                rb->offset = align;
-            }
-        }
```
Does this special case no longer apply?

mattcaswell replied:
This memmove was originally intended as an optimisation. It ensures that the payload data bytes are properly aligned for maximum assembler performance. There are two cases to consider: one where clearold is 0 and one where clearold is 1.

In the case that clearold is 0 this means "don't invalidate data earlier in the buffer because we're still using it". So this memove is just wrong in that instance. We must not move the data to an earlier position in the buffer because we're still using it. This actually was causing a problem in the pipelining case because we have multiple pipelines of data all using the same buffer - so this memmove was resulting in "bad record mac" error.

In the case that clearold is 1 it would still be a valid optimisation - but is entirely redundant  because of the memmove that occurs a few lines later. This *always* occurs if clearold is 1 and the data is not already aligned.

So, since this code is either wrong or redundant (depending on the value of clearold) it can safely be removed completely.

tmshort reviewed (approved):
Thanks for the explanations @mattcaswell

t8m commented:
Merged to master branch. Thank you.

mattcaswell commented:
> Merged to master branch. Thank you.

You beat me to it! I was just about to "git push"!
